source/meshworker/scratch_data.cc

```diff
@@ -0,0 +1,363 @@
+    const UpdateFlags &                 face_update_flags)
+    : ScratchData(StaticMappingQ1<dim, spacedim>::mapping,
+                  fe,
+                  quadrature,
+                  update_flags,
+                  face_quadrature,
+                  face_update_flags)
+  {}
+
+
+
+  template <int dim, int spacedim>
+  ScratchData<dim, spacedim>::ScratchData(
+    const FiniteElement<dim, spacedim> &fe,
+    const Quadrature<dim> &             quadrature,
+    const UpdateFlags &                 update_flags,
+    const UpdateFlags &                 neighbor_update_flags,
+    const Quadrature<dim - 1> &         face_quadrature,
+    const UpdateFlags &                 face_update_flags,
+    const UpdateFlags &                 neighbor_face_update_flags)
+    : ScratchData(StaticMappingQ1<dim, spacedim>::mapping,
+                  fe,
+                  quadrature,
+                  update_flags,
+                  neighbor_update_flags,
+                  face_quadrature,
+                  face_update_flags,
+                  neighbor_face_update_flags)
+  {}
+
+
+
+  template <int dim, int spacedim>
+  ScratchData<dim, spacedim>::ScratchData(
+    const ScratchData<dim, spacedim> &scratch)
+    : mapping(scratch.mapping)
+    , fe(scratch.fe)
+    , cell_quad(scratch.cell_quad)
+    , face_quad(scratch.face_quad)
+    , cell_flags(scratch.cell_flags)
+    , neighbor_cell_flags(scratch.neighbor_cell_flags)
+    , face_flags(scratch.face_flags)
+    , neighbor_face_flags(scratch.neighbor_face_flags)
+    , local_dof_indices(scratch.local_dof_indices)
+    , neighbor_dof_indices(scratch.neighbor_dof_indices)
+    , any_data(scratch.any_data)
+  {}
+
+
+
+  template <int dim, int spacedim>
+  const FEValues<dim, spacedim> &
+  ScratchData<dim, spacedim>::reinit(
+    const typename DoFHandler<dim, spacedim>::active_cell_iterator &cell)
+  {
+    if (!fe_values)
+      fe_values = std::make_unique<FEValues<dim, spacedim>>(*mapping,
+                                                            *fe,
+                                                            cell_quad,
+                                                            cell_flags);
+
+    fe_values->reinit(cell);
+    cell->get_dof_indices(local_dof_indices);
+    add_ref<FEValuesBase<dim, spacedim>>(*fe_values, "FEValuesBase");
+    return *fe_values;
+  }
+
+
+
+  template <int dim, int spacedim>
+  const FEFaceValues<dim, spacedim> &
+  ScratchData<dim, spacedim>::reinit(
+    const typename DoFHandler<dim, spacedim>::active_cell_iterator &cell,
+    const unsigned int                                              face_no)
+  {
+    if (!fe_face_values)
+      fe_face_values = std::make_unique<FEFaceValues<dim, spacedim>>(
+        *mapping, *fe, face_quad, face_flags);
+
+    fe_face_values->reinit(cell, face_no);
+    cell->get_dof_indices(local_dof_indices);
+    add_ref<FEValuesBase<dim, spacedim>>(*fe_face_values, "FEValuesBase");
```

Comment:
May I also ask, what is the rationale behind overwriting whatever is stored in the `FEValuesBase` slot? Would it not be possible to store this with the name `FEFaceValues` and return with `get_current_fe_face_values()`? I do understand why you would not want an additional entry for `FESubFaceValues` though (this facilitates the integration of jump terms like in [step-30](https://www.dealii.org/current/doxygen/deal.II/step_30.html#Functionassemble_system2), right?).

Reply:
I want to use the same slot for all FEValues/FaceValues/FESubfaceValues so that I can access them uniformly as get_current_fe, independently of whether they are on cell, face, subface, so that  get_current_fe_values() returns whatever you called reinit on.

Reply:
In this way `get_values` can be used both for cells, faces, and subfaces.

Reply:
I've moved FEValuesBase objects outside the boost::anydata map. This is faster and safer, and avoid the pitfall of having someone possibly overwriting the "FEValuesBase" slot.

Reply:
> In this way get_values can be used both for cells, faces, and subfaces.

Gotcha. It was this point that I failed to notice. Thanks for pointing that out!

Reply:
> I've moved FEValuesBase objects outside the boost::anydata map. This is faster and safer, and avoid the pitfall of having someone possibly overwriting the "FEValuesBase" slot.

Thanks for doing that :-)